src/executor/userOpMonitor.ts

```diff
@@ -0,0 +1,481 @@
+        return pendingBundles
+    }
+
+    async refreshBundleStatus(submittedBundle: SubmittedBundleInfo) {
+        let bundleReceipt = await getBundleStatus({
+            submittedBundle,
+            publicClient: this.config.publicClient,
+            logger: this.logger
+        })
+
+        if (bundleReceipt.status === "included") {
+            const { bundle } = submittedBundle
+            const { userOps, entryPoint } = bundle
+            const { transactionHash, blockNumber, userOpReceipts } =
+                bundleReceipt
+
+            // Cleanup bundle
+            await this.freeSubmittedBundle(submittedBundle)
+
+            // Log metric
+            this.metrics.userOperationsOnChain
+                .labels({ status: "included" })
+                .inc(userOps.length)
+
+            // Process each userOp
+            await Promise.all(
+                userOps.map(async (userOpInfo) => {
+                    const userOpReceipt = userOpReceipts[userOpInfo.userOpHash]
+                    if (!userOpReceipt) {
+                        throw new Error("userOpReceipt is undefined")
+                    }
+
+                    // Cache the receipt
+                    this.cacheReceipt(userOpInfo.userOpHash, userOpReceipt)
+
+                    await this.processIncludedUserOp(
+                        userOpInfo,
+                        userOpReceipt,
+                        transactionHash,
+                        blockNumber,
+                        entryPoint
+                    )
+                })
+            )
+        }
+
+        if (bundleReceipt.status === "reverted") {
+            const { bundle } = submittedBundle
+            const { userOps } = bundle
+            const { blockNumber, transactionHash } = bundleReceipt
+
+            // Cleanup bundle
+            await this.freeSubmittedBundle(submittedBundle)
+
+            // Fire and forget
+            Promise.all(
+                userOps.map(async (userOpInfo) => {
+                    this.checkFrontrun({
+                        userOpInfo,
+                        transactionHash,
+                        blockNumber
+                    })
+                })
+            )
+        }
+    }
+
+    public trackBundle(submittedBundle: SubmittedBundleInfo) {
+        const executor = submittedBundle.executor.address
+        this.pendingBundles.set(executor, submittedBundle)
+    }
+
+    // Helpers //
+    async getLatestBlockWithCache(): Promise<bigint> {
+        const now = Date.now()
+        const cache = this.cachedLatestBlock
+
+        if (cache && now - cache.timestamp < this.config.blockNumberCacheTtl) {
+            return cache.value
+        }
+
+        const latestBlock = await this.config.publicClient.getBlockNumber()
+        this.cachedLatestBlock = { value: latestBlock, timestamp: now }
+        return latestBlock
+    }
+
+    private cacheReceipt(userOpHash: Hex, receipt: any) {
+        this.cleanupReceiptCache()
```

Comment:
we should either cleanupReceiptCache on fetch or set, not both the times. I do not see the benefit on doing this twice. Also as stated can we make cleanup an async function